data/src/main/java/com/microsoft/azure/kusto/data/auth/UserPromptTokenProvider.java

```diff
@@ -0,0 +1,90 @@
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+public class UserPromptTokenProvider extends PublicAppTokenProviderBase {
+    private static final int USER_PROMPT_TIMEOUT_MS = 120 * 1000;
+    private static final URI redirectUri;
+
+    static {
+        URI tmp;
+        try {
+            tmp = new URI("http://localhost");
+        } catch (URISyntaxException e) { // This cannot happen, but is necessary to allow for a static final variable whose instantiation can throw an exception
+            tmp = null;
+        }
+        redirectUri = tmp;
+    }
+
+    private final String usernameHint;
+
+    UserPromptTokenProvider(@NotNull String clusterUrl, String authorityId) throws URISyntaxException {
+        this(null, clusterUrl, authorityId);
+    }
+
+    UserPromptTokenProvider(String usernameHint, @NotNull String clusterUrl, String authorityId) throws URISyntaxException {
+        super(clusterUrl, authorityId);
+        this.usernameHint = usernameHint;
+
+        try {
+            clientApplication = PublicClientApplication.builder(CLIENT_ID).authority(aadAuthorityUrl).build();
+        } catch (MalformedURLException e) {
+            throw new URISyntaxException(aadAuthorityUrl, ERROR_INVALID_AUTHORITY_URL);
+        }
+    }
+
+    @Override
+    protected IAuthenticationResult acquireNewAccessToken() throws DataServiceException, DataClientException {
+        IAuthenticationResult result;
+        try {
+            // This is the only auth method that allows the same application to be used for multiple distinct accounts, so reset account cache between sign-ins
+            clientApplication = PublicClientApplication.builder(CLIENT_ID).authority(aadAuthorityUrl).build();
+            CompletableFuture<IAuthenticationResult> future = clientApplication.acquireToken(InteractiveRequestParameters.builder(redirectUri).scopes(scopes).loginHint(usernameHint).build());
+            result = future.get(USER_PROMPT_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+        } catch (MalformedURLException e) {
+            throw new DataClientException(clusterUrl, ERROR_INVALID_AUTHORITY_URL, e);
+        } catch (TimeoutException | ExecutionException e) {
+            throw new DataServiceException(clusterUrl, ERROR_ACQUIRING_APPLICATION_ACCESS_TOKEN, e);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new DataServiceException(clusterUrl, ERROR_ACQUIRING_APPLICATION_ACCESS_TOKEN, e);
+        }
+        if (result == null) {
+            throw new DataServiceException(clusterUrl, "acquireWithUserPrompt got 'null' authentication result");
+        }
+        return result;
+    }
+
+    @Override
+    IAccount getAccount(Set<IAccount> accountSet) {

```

Comment:
Merge with the base API and accept a name that can be defaulted to null

Reply:
The username is indeed a property only relevant to UserPromptAuth. It's a different implementation.

Reply:
I still vote for same impl. with option for null hint